crates/worker/src/partition/state_machine/mod.rs

```diff
@@ -430,148 +423,177 @@ impl<Codec: RawEntryCodec> StateMachine<Codec> {
+                invocation_id
+            );
+            return Ok(Some(service_invocation));
+        }
+
+        if is_workflow_run && has_idempotency_key {
+            warn!("The idempotency key for workflow methods is ignored!");
+            has_idempotency_key = false;
+        }
+
+        let previous_invocation_status = async {
+            let mut is = ctx.get_invocation_status(&invocation_id).await?;
+            if is != InvocationStatus::Free {
+                // Deduplicated invocation with the new deterministic invocation id
+              Ok::<_, Error>(is)
             } else {
-                if self
-                    .try_resolve_idempotent_request(
-                        ctx,
-                        &idempotency_id,
-                        &service_invocation.invocation_id,
-                        service_invocation.response_sink.as_ref(),
-                    )
-                    .await?
-                    .is_some()
-                {
-                    Self::send_submit_notification_if_needed(
-                        ctx,
-                        service_invocation.invocation_id,
-                        false,
-                        service_invocation.submit_notification_sink,
-                    );
-                    debug_if_leader!(
-                        ctx.is_leader,
-                        restate.idempotency.id = ?idempotency_id,
-                        "Invocation is a duplicate"
-                    );
+                // We might still need to deduplicate based on the idempotency table for old invocation ids
+                if has_idempotency_key {
+                    let idempotency_id = service_invocation
+                        .compute_idempotency_id()
+                        .expect("Idempotency key must be present");
+
+                    if let Some(idempotency_metadata) = ctx.storage.get_idempotency_metadata(&idempotency_id).await? {
+                        is = ctx.get_invocation_status(&idempotency_metadata.invocation_id).await?;
+                    }
+                }
+                // Or on lock status for workflow runs with old invocation ids
+                if is_workflow_run {
+                    let keyed_service_id = service_invocation
+                        .invocation_target
+                        .as_keyed_service_id()
+                        .expect("When the handler type is Workflow, the invocation target must have a key");
 
-                    // Invocation was either resolved, or the sink was enqueued. Nothing else to do here.
-                    return Ok(None);
+                    if let VirtualObjectStatus::Locked(locked_invocation_id) = ctx
+                        .storage
+                        .get_virtual_object_status(&keyed_service_id)
+                        .await? {
+                        is = ctx.get_invocation_status(&locked_invocation_id).await?;
+                    }
                 }
+                Ok(is)
```

Comment:
Can we mark the code that will no longer be needed once we drop the old idempotency table?